File: TUnit.TestProject/Inject_SharedInstancePerKey.cs
using System.Collections.Concurrent;
using TUnit.TestProject.Attributes;

namespace TUnit.TestProject;

public static class SharedInjectedKeyedContainer
{
    public static readonly ConcurrentDictionary<string, List<DummyReferenceTypeClass>> InstancesPerKey = new();

    public static async Task Check(string key, DummyReferenceTypeClass dummyReferenceTypeClass)
    {
        var list = InstancesPerKey.GetOrAdd(key, _ => []);

        list.Add(dummyReferenceTypeClass);

        await Assert.That(list.Distinct()).HasSingleItem();
    }
}

[EngineTest(ExpectedResult.Pass)]
[ClassDataSource<DummyReferenceTypeClass>(Shared = SharedType.PerClass), NotInParallel]
public class InjectSharedPerKey1(DummyReferenceTypeClass dummyReferenceTypeClass)
{
    public static string Key
    {
        get
        {
            if (TestContext.Current is null)
            {
                throw new InvalidOperationException("TestContext.Current is null. This can happen if the test is not run in a TUnit test environment.");
            }
            return TestContext.Current!.TestDetails.ClassMetadata.Namespace + "." + TestContext.Current.TestDetails.ClassMetadata.Name + "_" + TestContext.Current.TestDetails.TestName;
        }
    }

    [Test, Repeat(5)]
    public async Task Test1()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }

    [Test, Repeat(5)]
    public async Task Test2()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }

    [Test, Repeat(5)]
    public async Task Test3()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }
}

[ClassDataSource<DummyReferenceTypeClass>(Shared = SharedType.PerClass), NotInParallel]
public class InjectSharedPerKey2(DummyReferenceTypeClass dummyReferenceTypeClass)
{
    public static string Key => TestContext.Current!.TestDetails.ClassMetadata.Namespace + "." + TestContext.Current.TestDetails.ClassMetadata.Name  + "_" + TestContext.Current.TestDetails.TestName;

    [Test, Repeat(5)]
    public async Task Test1()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }

    [Test, Repeat(5)]
    public async Task Test2()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }

    [Test, Repeat(5)]
    public async Task Test3()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }
}

[ClassDataSource<DummyReferenceTypeClass>(Shared = SharedType.PerClass), NotInParallel]
public class InjectSharedPerKey3(DummyReferenceTypeClass dummyReferenceTypeClass)
{
    public static string Key => TestContext.Current!.TestDetails.ClassMetadata.Namespace + "." + TestContext.Current.TestDetails.ClassMetadata.Name + "_"
        + TestContext.Current.TestDetails.TestName;

    [Test, Repeat(5)]
    public async Task Test1()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }

    [Test, Repeat(5)]
    public async Task Test2()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }

    [Test, Repeat(5)]
    public async Task Test3()
    {
        await SharedInjectedKeyedContainer.Check(Key, dummyReferenceTypeClass);
    }
}

File: TUnit.TestProject.FSharp/Inject_SharedInstancePerKey.fs
namespace TUnit.TestProject.FSharp

open System.Collections.Concurrent
open System.Linq
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Core

// F# equivalent of SharedInjectedKeyedContainer
module SharedInjectedKeyedContainer =
    let instancesPerKey = ConcurrentDictionary<string, ResizeArray<DummyReferenceTypeClass>>()

type DummyReferenceTypeClass() = class end

[<ClassDataSource(typeof<DummyReferenceTypeClass>, Shared=SharedType.PerClass)>]
[<NotInParallel>]
type InjectSharedPerKey1(dummyReferenceTypeClass: DummyReferenceTypeClass) =
    [<Test>]
    [<Repeat(5)>]
    member _.Test1() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }
    [<Test>]
    [<Repeat(5)>]
    member _.Test2() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }
    [<Test>]
    [<Repeat(5)>]
    member _.Test3() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }

[<ClassDataSource(typeof<DummyReferenceTypeClass>, Shared=SharedType.PerClass)>]
[<NotInParallel>]
type InjectSharedPerKey2(dummyReferenceTypeClass: DummyReferenceTypeClass) =
    [<Test>]
    [<Repeat(5)>]
    member _.Test1() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }
    [<Test>]
    [<Repeat(5)>]
    member _.Test2() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }
    [<Test>]
    [<Repeat(5)>]
    member _.Test3() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }

[<ClassDataSource(typeof<DummyReferenceTypeClass>, Shared=SharedType.PerClass)>]
[<NotInParallel>]
type InjectSharedPerKey3(dummyReferenceTypeClass: DummyReferenceTypeClass) =
    [<Test>]
    [<Repeat(5)>]
    member _.Test1() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }
    [<Test>]
    [<Repeat(5)>]
    member _.Test2() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }
    [<Test>]
    [<Repeat(5)>]
    member _.Test3() = task {
        let testName = TestContext.Current.Value.TestDetails.TestName
        let found, list = SharedInjectedKeyedContainer.instancesPerKey.TryGetValue(testName)
        if found && list.Count > 0 then
            do! Assert.That(list).Contains(dummyReferenceTypeClass)
        for KeyValue(k, v) in SharedInjectedKeyedContainer.instancesPerKey do
            if k <> testName then
                do! Assert.That(list).DoesNotContain(dummyReferenceTypeClass)
        let l = SharedInjectedKeyedContainer.instancesPerKey.GetOrAdd(testName, fun _ -> ResizeArray())
        l.Add(dummyReferenceTypeClass)
        do! Assert.That(l.Distinct()).HasSingleItem()
    }